flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultSchedulerFactory.java

```diff
@@ -125,24 +153,50 @@ static SchedulingStrategyFactory createSchedulingStrategyFactory(final ScheduleM
 		}
 	}
 
-	private static ExecutionSlotAllocatorFactory createExecutionSlotAllocatorFactory(
+	private static DefaultSchedulerComponents createPipelinedRegionSchedulerComponents(
 			final ScheduleMode scheduleMode,
-			final SlotProvider slotProvider,
-			final Time slotRequestTimeout,
-			final SchedulingStrategyFactory schedulingStrategyFactory) {
-
-		if (schedulingStrategyFactory instanceof PipelinedRegionSchedulingStrategy.Factory) {
-			return new OneSlotPerExecutionSlotAllocatorFactory(
-				slotProvider,
-				scheduleMode != ScheduleMode.LAZY_FROM_SOURCES_WITH_BATCH_SLOT_REQUEST,
-				slotRequestTimeout);
-		} else {
-			final SlotProviderStrategy slotProviderStrategy = SlotProviderStrategy.from(
-				scheduleMode,
-				slotProvider,
-				slotRequestTimeout);
-
-			return new DefaultExecutionSlotAllocatorFactory(slotProviderStrategy);
+			final Configuration jobMasterConfiguration,
+			final SlotPool slotPool,
+			final Time slotRequestTimeout) {
+		final SlotSelectionStrategy slotSelectionStrategy = selectSlotSelectionStrategy(jobMasterConfiguration);
+		final BulkSlotProvider bulkSlotProvider = new BulkSlotProviderImpl(slotSelectionStrategy, slotPool);
+		final ExecutionSlotAllocatorFactory allocatorFactory = new OneSlotPerExecutionSlotAllocatorFactory(
+			bulkSlotProvider,
+			scheduleMode != ScheduleMode.LAZY_FROM_SOURCES_WITH_BATCH_SLOT_REQUEST,
+			slotRequestTimeout);
+		return new DefaultSchedulerComponents(
+			new PipelinedRegionSchedulingStrategy.Factory(),
+			bulkSlotProvider::start,
+			allocatorFactory);
+	}
+
+	@Nonnull
```

Comment:
We can remove the `@Nonnull` annotation I think.
This is because that we will by default assume it to be non-null, 
and only `@Nullable` annotations are required when needed.